Feature explain analyze v2

sopel39 opened the pull request.

sopel39 commented:
Supersedes: #5291

Squashed, rebased, applied comments
FYI: @cberner

cberner commented:
I merged the first two commits. Can you move the rewording of "unknown" -> "?" to a separate PR/issue for discussion? Don't want to hold up this one with that discussion

sopel39 commented:
@cberner I would prefer to make a separate commit with `unknown -> ?` change before: https://github.com/prestodb/presto/pull/5499/commits/314bd1740d929f693aaf364b96297daa7332bc8b. I don't want to hold https://github.com/prestodb/presto/pull/5499/commits/d6ceaf1651d73ba8cc10634ca2a75ca6b34eed18. I think it would be valuable that you review https://github.com/prestodb/presto/pull/5499/commits/d6ceaf1651d73ba8cc10634ca2a75ca6b34eed18 too so that you understand why I think that `?` makes output more readable. https://github.com/prestodb/presto/pull/5499/commits/d6ceaf1651d73ba8cc10634ca2a75ca6b34eed18 adds advanced new stats which become verbose if we use `unknown` in each place where the value is not available. For example:
```
PartitionedOutputOperator := Collisions avg.: ? (?% est.), Collisions std.dev.: ?%.
```
vs
```
PartitionedOutputOperator := Collisions avg.: unknown (unknown% est.), Collisions std.dev.: unknown%.
```
If you think such lines are OK then I will leave `unknown` in all places and remove it from PR.
What do you think?

cberner commented:
@dain and @electrum are the ones we need to talk to about "unknown" vs "?" actually. They're at Hadoop Summit at the moment, and I'm on vacation the long US weekend, but I can discuss with them next week, or you can message them directly

sopel39 commented:
I have extracted `unknown -> ?` into https://github.com/prestodb/presto/pull/5499/commits/75fa108b5e5fceaef7b0221c83053f7213227c8e so that https://github.com/prestodb/presto/pull/5499/commits/b3b676d5e01e37f24514f24730a21af8047f15fe can be reviewed first (it is before https://github.com/prestodb/presto/pull/5499/commits/75fa108b5e5fceaef7b0221c83053f7213227c8e)

BTW: have there been an improvement by @nileema how operator stats are collected? I haven't noticed any `unknowns` in short `EXPLAIN ANALYZE` queries.

sopel39 commented:
rebased ping! @cberner

cberner commented:
Oh sorry, didn't realize this was ready for review. Please assign PRs back to me when they're ready, as that's how I check what I need to be reviewing

sopel39 commented:
I have rebased on newest master. Would you like to take a look again @cberner ?

cberner commented:
Ya, happy to take a look. Could you sync up with @losipiuk and others about the priority of Teradata PRs I have assigned to me? He asked that I look at the revokable memory stuff next, and first I'm finishing up the `char(x)` PR

cberner commented:
Sorry about the delay here, but I've finally finished with the spilling PR and looks like this one is next on the priority list. Can you rebase to resolve the conflicts and then assign back to me?

sopel39 commented:
Sure, thanks. I will do that.

sopel39 commented:
updated & rebased

cberner commented:
First two commits look good. Still looking at the rest

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java:
```diff
@@ -176,10 +179,9 @@ public void finish()
+        LookupSourceSupplier partition = index.createLookupSourceSupplier(operatorContext.getSession(), hashChannels, preComputedHashChannel, filterFunctionFactory);
         lookupSourceFactory.setPartitionLookupSourceSupplier(partitionIndex, partition);
-
-        operatorContext.setMemoryReservation(partition.get().getInMemorySizeInBytes());
```
Why did you remove this?

sopel39 replied:
Sorry, my bad. Must have been removed during conflict resolution.

cberner commented on presto-main/src/main/java/com/facebook/presto/util/HashCollisionsEstimator.java:
```diff
@@ -0,0 +1,83 @@
+import static java.lang.Math.ceil;
+import static java.lang.Math.floor;
+
+public final class HashCollisionsEstimator
```
I think we need to find a closed form expression for this. 10k values is much too small, as today we could easily have 10B+. Have you looked up the Birthday Paradox problem? I would expect this to be well studied

sopel39 replied:
Those estimates scale accurately to larger hash tables. For instance assume you have estimated number of collisions `c` for inserting `x` entries into `y` sized hash table. Then the number of collisions for inserting `x*scale` entries into `y*scale` sized hash table is exactly `c*scale`. You can estimate the number of collisions in such way for hash table of any size. Your relative error will remain the same independently of table size. The greater number of estimates `N`, the better the estimate. However, I have found that with `500` data points I get really accurate (less than `1%` error) (and useful) results in explain analyze output. 

It's much easier to deduce closed formula for Birthday problem and hash tables where after collision you uniformly try next pos (instead of trying the next table cell linearly). In fact, I deduced formula for the latter one.

It is not possible though to deduce closed formula for hash tables where after collision you try the next table cell linearly. This is because you cannot assume uniform distribution of table entries when inserting new element (each recursion step is different). We though about closed formula with my mathematician colleague, but concluded that it is rather not possible (did not hard prove the fact through).

Therefore I propose we keep it estimated since it is a really nice feature (e.g: we've seem performance improvements due to hash computation optimizations by @martint) and allows user to identify potential data or hashing problems.

cberner replied:
I see. Could you add some comments to this class that describe why this linearization works, and point to the class used to generate the constants? It looks like they were only generated for up to 10k values, and it's not clear to be that we can extrapolate linearly out to say 10 trillion values

sopel39 replied:
Okey. I have added comments and empirical test to confirm that hash collisions scales almost linearly (see TestHashCollisionsEstimator#hashEstimatesShouldScale)

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java:
```diff
@@ -211,11 +220,18 @@ public static String textDistributedPlan(List<StageInfo> stages, Metadata metada
+        Map<PlanNodeId, Long> planNodeOutputBytes = new HashMap<>();
+        Map<PlanNodeId, Long> planNodeWallMillis = new HashMap<>();
+
+        Map<PlanNodeId, Map<String, Long>> operatorTotalDrivers = new HashMap<>();
```
Why are these named "operator" instead of "planNode"

sopel39 replied:
In order to compute collisions or input standard deviation, I need to compute stats for each operator that was generated for given `PlanNode`. For instance:
```
- InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_7] => [orderkey:bigint, $hashvalue:bigint, orderkey_0:bigint, $hashvalue_7:bigint] 
        Cost: 32.52%, Output: 6001215 rows (206.04MB)                                                                                              
        LookupJoinOperator := Drivers: 4, Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                      
        HashBuilderOperator := Drivers: 16, Input avg.: 93750.00 lines, Input std.dev.: 0.23%                                                      
        HashBuilderOperator := Collisions avg.: 117186.29 (99.60% est.), Collisions std.dev.: 1.56%
```
There are different number of drivers for `LookupJoinOperator` and `HashBuilderOperator`. Also, only `HashBuilderOperator` provides collisions data

cberner replied:
When is it helpful to know the total number of drivers? I think it would be simpler to print this more like:
```
- InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_7] => [orderkey:bigint, $hashvalue:bigint, orderkey_0:bigint, $hashvalue_7:bigint] 
        Cost: 32.52%, Output: 6001215 rows (206.04MB)                                                                                              
        Left (probe) input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                      
        Right (build) input avg.: 93750.00 lines, Input std.dev.: 0.23%                                                      
                Collisions avg.: 117186.29 (99.60% est.), Collisions std.dev.: 1.56%
```

sopel39 replied:
The number of drivers is useful when diagnosing concurrency and parallelism. It might quickly show some misconfiguration (task concurrency, multiple tasks per worker) or other issues.

I might remove it though if you want.

cberner replied:
I think a more useful metric would be thread/compute utilization. I don't think the average end user has a good understanding of what drivers are and how they work

sopel39 replied:
I have adjusted output as you suggested. Now it looks like:
```
             - InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_7] => [orderkey:bigint, $hashvalue:bigint, orderkey_0:bigint, $has
                     Cost: 39.15%, Output: 600572 rows (20.62MB)                                                                               
                     Left (probe) Input avg.: 150143.00 lines, Input std.dev.: 0.33%                                                           
                     Right (build) Input avg.: 9375.00 lines, Input std.dev.: 0.80%                                                            
                             Collisions avg.: 6244.50 (99.59% est.), Collisions std.dev.: 5.03%
```

However, I still wonder whether it makes sense to report `input avg`, `collisions avg` without the number of drivers that were running? I could just report `input std dev`, `collisions std dev`, but maybe it's better to display number of drivers with better name, say: `# of executors`?

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java:
```diff
@@ -1211,67 +1309,193 @@ private static String castToVarchar(Type type, Object value, Metadata metadata,
         }
     }
+
+    private static double computerStdDev(double sumSquared, double sum, long n)
```
typo

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java:
```diff
@@ -418,38 +481,73 @@ private void printStats(int indent, PlanNodeId planNodeId, boolean printInput, b
+
+    private void printDistributions(int indent, PlanNodeStats nodeStats)
+    {
+        Map<String, Long> totalDrivers = nodeStats.getOperatorTotalDrivers();
```
For now I don't think we should be printing information about operators. Is there a case where this adds a lot of value? For the most part they map 1:1 to plan nodes, and in the future that should be even more true

sopel39 replied:
Take a look at example output of explain analyze with operators stats reporting:
```
 Fragment 2 [tpch:orders:1500000]                                                                                                                               
     Cost: CPU 7.54s, Input: 7501215 rows (128.77MB), Output: 4 rows (36B)                                                                                      
     Output layout: [count_6]                                                                                                                                   
     Output partitioning: SINGLE []                                                                                                                             
     - Aggregate(PARTIAL) => [count_6:bigint]                                                                                                                   
             Cost: 0.02%, Output: 4 rows (36B)                                                                                                                  
             AggregationOperator := Drivers: 4, Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                             
             TaskOutputOperator := Drivers: 4, Input avg.: 1.00 lines, Input std.dev.: 0.00%                                                                    
             count_6 := "count"(*)                                                                                                                              
         - Project[] => []                                                                                                                                      
                 Cost: 0.38%, Input: 6001215 rows (206.04MB), Output: 6001215 rows (0B), Filtered: 0.00%                                                        
                 FilterAndProjectOperator := Drivers: 4, Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                    
             - InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_7] => [orderkey:bigint, $hashvalue:bigint, orderkey_0:bigint, $hashvalue_7:bigint] 
                     Cost: 36.17%, Output: 6001215 rows (206.04MB)                                                                                              
                     LookupJoinOperator := Drivers: 4, Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                      
                     HashBuilderOperator := Drivers: 16, Input avg.: 93750.00 lines, Input std.dev.: 0.23%                                                      
                                         := Collisions avg.: 117186.29 (99.60% est.), Collisions std.dev.: 1.56%                                                
                 - ScanProject[table = tpch:tpch:lineitem:sf1.0, originalConstraint = true] => [orderkey:bigint, $hashvalue:bigint]                             
                         Cost: 57.16%, Input: 6001215 rows (0B), Output: 6001215 rows (103.02MB), Filtered: 0.00%                                               
                         ScanFilterAndProjectOperator := Drivers: 4, Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                        
                         $hashvalue := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))                                               
                         orderkey := tpch:orderkey                                                                                                              
                 - LocalExchange[HASH][$hashvalue_7] ("orderkey_0") => orderkey_0:bigint, $hashvalue_7:bigint                                                   
                         Cost: 6.19%, Output: 1500000 rows (25.75MB)                                                                                            
                         LocalExchangeSourceOperator := Drivers: 16, Input avg.: 93750.00 lines, Input std.dev.: 0.23%                                          
                         LocalExchangeSinkOperator := Drivers: 1, Input avg.: 1500000.00 lines, Input std.dev.: 0.00%                                           
                     - RemoteSource[3] => [orderkey_0:bigint, $hashvalue_8:bigint]                                                                              
                             Cost: 0.07%, Output: 1500000 rows (25.75MB)                                                                                        
                             ExchangeOperator := Drivers: 1, Input avg.: 1500000.00 lines, Input std.dev.: 0.00%      
```

`Aggregate`, `InnerJoin` and `LocalExchange` are all composed of multiple operators (half of total plan nodes). The information about skewness and hash collisions is especially useful for us when working on remote client cluster where we have limited access. In such cases by looking at `EXPLAIN ANALYZE` with individual operator stats we can quickly deduce how the data flow looks like and where there might be a bottleneck due to the skewness or excessive collisions. Since we have the information, it should be beneficial to display it even though one might not use every time (but rather when tracking some unexpected performance issue). Trying to squash the stats will cause some information to be hidden, which might hide valuable clues.

cberner replied:
`TaskOutputOperator` is not really part of the aggregation. It's just there at the end of every pipeline. What problems have you run into where having all the operator info has helped? It seems like it clutters up the analysis, and most of the information could be rendered in a better way (see my comment above about the JOIN)

sopel39 commented:
Added code & docs cleanups. Back to you @cberner

cberner commented on presto-main/src/main/java/com/facebook/presto/util/HashCollisionsEstimator.java:
```diff
@@ -0,0 +1,83 @@
+        return (lowerEstimation * (1 - estimationIndexDistanceFromLower) + upperEstimation * estimationIndexDistanceFromLower) * estimateRescaleFactor;
+    }
+
+    private HashCollisionsEstimator() {}
```
Constructor should be at top of class

cberner commented:
I ran into a NPE testing this in our test tier. Unfortunately, I wasn't able to reproduce it with TPCH data, but the query is of this form:
```
select count(distinct x)
from t
where a = '2016-11-01'
and b = 'xxx'
and c = 'yyy'
and d = 'zzz' 
LIMIT 10000
```
where `t` is partitioned on `a`
and the stack trace is:
```
Query 20161202_232144_00317_7tgfd failed: null
java.lang.NullPointerException
	at com.facebook.presto.sql.planner.PlanPrinter.getPlanNodeStats(PlanPrinter.java:308)
	at com.facebook.presto.sql.planner.PlanPrinter.lambda$textDistributedPlan$1(PlanPrinter.java:204)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at com.facebook.presto.sql.planner.PlanPrinter.textDistributedPlan(PlanPrinter.java:205)
	at com.facebook.presto.operator.ExplainAnalyzeOperator.getOutput(ExplainAnalyzeOperator.java:149)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:378)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:301)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:622)
	at com.facebook.presto.execution.TaskExecutor$PrioritizedSplitRunner.process(TaskExecutor.java:534)
	at com.facebook.presto.execution.TaskExecutor$Runner.run(TaskExecutor.java:670)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

cberner commented:
Back to you with some more comments and a bug I found

sopel39 commented:
Could you give some more info about the bug, e.g: commit hash or copy-paste code line?

The line you pointed doesn't seem to be problematic in code, but my code might be misaligned.

cberner commented:
I used this branch unchanged, but rebased it on b757f4c4f553fa0bafb2f67146028e3a6e854507

cberner commented:
The NPE is in this for-loop:
```
List<PlanNodeStats> stats = new ArrayList<>();
        for (Map.Entry<PlanNodeId, Long> entry : planNodeWallMillis.entrySet()) {
            PlanNodeId planNodeId = entry.getKey();
            stats.add(new PlanNodeStats(
                    planNodeId,
                    new Duration(planNodeWallMillis.get(planNodeId), MILLISECONDS),
                    planNodeInputPositions.get(planNodeId),
                    succinctDataSize(planNodeInputBytes.get(planNodeId), BYTE),
                    // It's possible there will be no output stats because all the pipelines that we observed were non-output.
                    // For example in a query like SELECT * FROM a JOIN b ON c = d LIMIT 1
                    // It's possible to observe stats after the build starts, but before the probe does
                    // and therefore only have wall time, but no output stats
                    planNodeOutputPositions.getOrDefault(planNodeId, 0L),
                    succinctDataSize(planNodeOutputBytes.getOrDefault(planNodeId, 0L), BYTE),
                    operatorInputStats.get(planNodeId),
                    // Only some operators emit hash collisions statistics
                    operatorHashCollisionsStats.getOrDefault(planNodeId, emptyMap())));
        }
```

on this line: `planNodeInputPositions.get(planNodeId),`

sopel39 commented:
This line:
```
planNodeInputPositions.get(planNodeId)
```
shouldn't fail because of NPE. `planNodeInputPositions` is always initialized as `HashMap`, so even null key should work. Not sure what could be a problem there. What data did you use?

sopel39 commented:
@cberner I have applied comments. Now output looks like:
```
presto:sf1> EXPLAIN ANALYZE SELECT count(*), clerk FROM orders WHERE orderdate > date '1995-01-01' GROUP BY clerk;
                                                                                                                   Query Plan                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [HASH]                                                                                                                                                                                          
     Cost: CPU 38.82ms, Input: 4000 rows (148.44kB), Output: 1000 rows (28.32kB)                                                                                                                            
     Output layout: [count, clerk]                                                                                                                                                                          
     Output partitioning: SINGLE []                                                                                                                                                                         
     - Project[] => [count:bigint, clerk:varchar(15)]                                                                                                                                                       
             Cost: 5.49%, Input: 1000 rows (37.11kB), Output: 1000 rows (28.32kB), Filtered: 0.00%                                                                                                          
             Input avg.: 62.50 lines, Input std.dev.: 14.77%                                                                                                                                                
         - Aggregate(FINAL)[clerk][$hashvalue] => [clerk:varchar(15), $hashvalue:bigint, count:bigint]                                                                                                      
                 Cost: 46.15%, Output: 1000 rows (37.11kB)                                                                                                                                                  
                 Input avg.: 250.00 lines, Input std.dev.: 14.77%                                                                                                                                           
                 count := "count"("count_8")                                                                                                                                                                
             - LocalExchange[HASH][$hashvalue] ("clerk") => clerk:varchar(15), count_8:bigint, $hashvalue:bigint                                                                                            
                     Cost: 38.46%, Output: 4000 rows (148.44kB)                                                                                                                                             
                     Input avg.: 4000.00 lines, Input std.dev.: 0.00%                                                                                                                                       
                 - RemoteSource[2] => [clerk:varchar(15), count_8:bigint, $hashvalue_9:bigint]                                                                                                              
                         Cost: 9.89%, Output: 4000 rows (148.44kB)                                                                                                                                          
                         Input avg.: 4000.00 lines, Input std.dev.: 0.00%                                                                                                                                   
                                                                                                                                                                                                            
 Fragment 2 [tpch:orders:1500000]                                                                                                                                                                           
     Cost: CPU 7.00s, Input: 818058 rows (22.62MB), Output: 4000 rows (148.44kB)                                                                                                                            
     Output layout: [clerk, count_8, $hashvalue_10]                                                                                                                                                         
     Output partitioning: HASH [clerk][$hashvalue_10]                                                                                                                                                       
     - Aggregate(PARTIAL)[clerk][$hashvalue_10] => [clerk:varchar(15), $hashvalue_10:bigint, count_8:bigint]                                                                                                
             Cost: 2.16%, Output: 4000 rows (148.44kB)                                                                                                                                                      
             Input avg.: 204514.50 lines, Input std.dev.: 0.05%                                                                                                                                             
             Collisions avg.: 5701.28 (17569.93% est.), Collisions std.dev.: 1.12%                                                                                                                          
             count_8 := "count"(*)                                                                                                                                                                          
         - ScanFilterProject[table = tpch:tpch:orders:sf1.0, originalConstraint = ("orderdate" > "$literal$date"(BIGINT '9131')), filterPredicate = ("orderdate" > "$literal$date"(BIGINT '9131'))] => [cler
                 Cost: 97.84%, Input: 1500000 rows (0B), Output: 818058 rows (22.62MB), Filtered: 45.46%                                                                                                    
                 Input avg.: 375000.00 lines, Input std.dev.: 0.00%                                                                                                                                         
                 $hashvalue_10 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("clerk"), 0))                                                                                                   
                 orderdate := tpch:orderdate                                                                                                                                                                
                 clerk := tpch:clerk
```
```
presto:sf1> explain analyze select count(*) from lineitem l, orders o where o.orderkey = l.orderkey;
                                                                           Query Plan                                                                           
----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [SINGLE]                                                                                                                                            
     Cost: CPU 3.59ms, Input: 4 rows (36B), Output: 1 row (9B)                                                                                                  
     Output layout: [count]                                                                                                                                     
     Output partitioning: SINGLE []                                                                                                                             
     - Aggregate(FINAL) => [count:bigint]                                                                                                                       
             Cost: 7.61%, Output: 1 row (9B)                                                                                                                    
             Input avg.: 4.00 lines, Input std.dev.: 0.00%                                                                                                      
             count := "count"("count_6")                                                                                                                        
         - RemoteSource[2] => [count_6:bigint]                                                                                                                  
                 Cost: 92.39%, Output: 4 rows (36B)                                                                                                             
                 Input avg.: 4.00 lines, Input std.dev.: 0.00%                                                                                                  
                                                                                                                                                                
 Fragment 2 [tpch:orders:1500000]                                                                                                                               
     Cost: CPU 11.32s, Input: 7501215 rows (128.77MB), Output: 4 rows (36B)                                                                                     
     Output layout: [count_6]                                                                                                                                   
     Output partitioning: SINGLE []                                                                                                                             
     - Aggregate(PARTIAL) => [count_6:bigint]                                                                                                                   
             Cost: 5.08%, Output: 4 rows (36B)                                                                                                                  
             Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                                                                
             count_6 := "count"(*)                                                                                                                              
         - Project[] => []                                                                                                                                      
                 Cost: 0.39%, Input: 6001215 rows (206.04MB), Output: 6001215 rows (0B), Filtered: 0.00%                                                        
                 Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                                                            
             - InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_7] => [orderkey:bigint, $hashvalue:bigint, orderkey_0:bigint, $hashvalue_7:bigint] 
                     Cost: 34.68%, Output: 6001215 rows (206.04MB)                                                                                              
                     Left (probe) Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                                           
                     Right (build) Input avg.: 93750.00 lines, Input std.dev.: 0.23%                                                                            
                             Collisions avg.: 117186.29 (99.60% est.), Collisions std.dev.: 1.56%                                                               
                 - ScanProject[table = tpch:tpch:lineitem:sf1.0, originalConstraint = true] => [orderkey:bigint, $hashvalue:bigint]                             
                         Cost: 53.03%, Input: 6001215 rows (0B), Output: 6001215 rows (103.02MB), Filtered: 0.00%                                               
                         Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                                                    
                         $hashvalue := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))                                               
                         orderkey := tpch:orderkey                                                                                                              
                 - LocalExchange[HASH][$hashvalue_7] ("orderkey_0") => orderkey_0:bigint, $hashvalue_7:bigint                                                   
                         Cost: 6.76%, Output: 1500000 rows (25.75MB)                                                                                            
                         Input avg.: 1500000.00 lines, Input std.dev.: 0.00%                                                                                    
                     - RemoteSource[3] => [orderkey_0:bigint, $hashvalue_8:bigint]                                                                              
                             Cost: 0.06%, Output: 1500000 rows (25.75MB)                                                                                        
                             Input avg.: 1500000.00 lines, Input std.dev.: 0.00%                                                                                
                                                                                                                                                                
 Fragment 3 [tpch:orders:1500000]                                                                                                                               
     Cost: CPU 11.17s, Input: 1500000 rows (25.75MB), Output: 1500000 rows (25.75MB)                                                                            
     Output layout: [orderkey_0, $hashvalue_9]                                                                                                                  
     Output partitioning: tpch:orders:1500000 [orderkey_0]                                                                                                      
     - ScanProject[table = tpch:tpch:orders:sf1.0, originalConstraint = true] => [orderkey_0:bigint, $hashvalue_9:bigint]                                       
             Cost: 100.00%, Input: 1500000 rows (0B), Output: 1500000 rows (25.75MB), Filtered: 0.00%                                                           
             Input avg.: 375000.00 lines, Input std.dev.: 0.00%                                                                                                 
             $hashvalue_9 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey_0"), 0))                                                       
             orderkey_0 := tpch:orderkey  
```

I couldn't reproduce the bug you mentioned. The exception might be due some partial stalled compilation (we've seen some strange exceptions due to this cause before). Otherwise the line you pointed shouldn't cause NPE in any circumstances (the `HashMap` is always initialized).

cberner commented:
I tried to rebase on master, but there were conflicts. Could you rebase and then I'll test it again and try to provide more information on the NPE?

cberner commented:
Oh, I just didn't fetch your latest changes, nm. I'll test it now

cberner commented:
Ah, I think my local branch was just out of date. I fetched your latest changes and rebased on 03635fec5001f3f102b0a3f1460d589cc274ce2f and now it works fine

sopel39 commented:
@cberner Have you had a change to work on this PR?

cberner commented:
Unfortunately, I haven't. I've been working on getting the 0.161 release out, and then am leaving on vacation at the end of the week, so likely won't be able to look at this until mid-January

sopel39 commented:
rebased @cberner

cberner commented:
Merged "Improvements to explain analyze" and "Use ? instead of "unknown" in EXPLAIN ANALYZE". Still looking at the others

cberner commented:
@sopel39 how many cases are there where there's more than one `Operator` for a `PlanNode`? I think JOINs are the only case, in which case I think we should render those operator level stats as part of the `PlanNode` rather than showing them separately.

sopel39 commented:
Squashed fixups

@cberner please have a look at one of previous comment: https://github.com/prestodb/presto/pull/5499#discussion_r90444071. There are various plan nodes that are composed of multiple operators (e.g: `InnerJoin`, `LocalExchange`).

Anyway, I have addressed your comment (I squashed that fixup) and now plan is more verbose as in comment: https://github.com/prestodb/presto/pull/5499#issuecomment-265082139. See method: `PlanPrinter#printDistributions`. It only displays operator stats when meaningful (e.g: there are collisions reported)

Example output: 
```
presto:sf1> EXPLAIN ANALYZE SELECT count(*), clerk FROM orders WHERE orderdate > date '1995-01-01' GROUP BY clerk;
                                                                                                                   Query Plan                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [HASH]                                                                                                                                                                                          
     Cost: CPU 38.82ms, Input: 4000 rows (148.44kB), Output: 1000 rows (28.32kB)                                                                                                                            
     Output layout: [count, clerk]                                                                                                                                                                          
     Output partitioning: SINGLE []                                                                                                                                                                         
     - Project[] => [count:bigint, clerk:varchar(15)]                                                                                                                                                       
             Cost: 5.49%, Input: 1000 rows (37.11kB), Output: 1000 rows (28.32kB), Filtered: 0.00%                                                                                                          
             Input avg.: 62.50 lines, Input std.dev.: 14.77%                                                                                                                                                
         - Aggregate(FINAL)[clerk][$hashvalue] => [clerk:varchar(15), $hashvalue:bigint, count:bigint]                                                                                                      
                 Cost: 46.15%, Output: 1000 rows (37.11kB)                                                                                                                                                  
                 Input avg.: 250.00 lines, Input std.dev.: 14.77%                                                                                                                                           
                 count := "count"("count_8")                                                                                                                                                                
             - LocalExchange[HASH][$hashvalue] ("clerk") => clerk:varchar(15), count_8:bigint, $hashvalue:bigint                                                                                            
                     Cost: 38.46%, Output: 4000 rows (148.44kB)                                                                                                                                             
                     Input avg.: 4000.00 lines, Input std.dev.: 0.00%                                                                                                                                       
                 - RemoteSource[2] => [clerk:varchar(15), count_8:bigint, $hashvalue_9:bigint]                                                                                                              
                         Cost: 9.89%, Output: 4000 rows (148.44kB)                                                                                                                                          
                         Input avg.: 4000.00 lines, Input std.dev.: 0.00%                                                                                                                                   
                                                                                                                                                                                                            
 Fragment 2 [tpch:orders:1500000]                                                                                                                                                                           
     Cost: CPU 7.00s, Input: 818058 rows (22.62MB), Output: 4000 rows (148.44kB)                                                                                                                            
     Output layout: [clerk, count_8, $hashvalue_10]                                                                                                                                                         
     Output partitioning: HASH [clerk][$hashvalue_10]                                                                                                                                                       
     - Aggregate(PARTIAL)[clerk][$hashvalue_10] => [clerk:varchar(15), $hashvalue_10:bigint, count_8:bigint]                                                                                                
             Cost: 2.16%, Output: 4000 rows (148.44kB)                                                                                                                                                      
             Input avg.: 204514.50 lines, Input std.dev.: 0.05%                                                                                                                                             
             Collisions avg.: 5701.28 (17569.93% est.), Collisions std.dev.: 1.12%                                                                                                                          
             count_8 := "count"(*)                                                                                                                                                                          
         - ScanFilterProject[table = tpch:tpch:orders:sf1.0, originalConstraint = ("orderdate" > "$literal$date"(BIGINT '9131')), filterPredicate = ("orderdate" > "$literal$date"(BIGINT '9131'))] => [cler
                 Cost: 97.84%, Input: 1500000 rows (0B), Output: 818058 rows (22.62MB), Filtered: 45.46%                                                                                                    
                 Input avg.: 375000.00 lines, Input std.dev.: 0.00%                                                                                                                                         
                 $hashvalue_10 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("clerk"), 0))                                                                                                   
                 orderdate := tpch:orderdate                                                                                                                                                                
                 clerk := tpch:clerk
```

Example join output:
```
             - InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_7] => [orderkey:bigint, $hashvalue:bigint, orderkey_0:bigint, $hashvalue_7:bigint] 
                     Cost: 34.68%, Output: 6001215 rows (206.04MB)                                                                                              
                     Left (probe) Input avg.: 1500303.75 lines, Input std.dev.: 0.03%                                                                           
                     Right (build) Input avg.: 93750.00 lines, Input std.dev.: 0.23%                                                                            
                             Collisions avg.: 117186.29 (99.60% est.), Collisions std.dev.: 1.56%                                                               

```

sopel39 commented:
Did you have time to look at new output @cberner?

cberner commented:
Cool, ya that looks good to me! Will take another look through the code now

sopel39 commented:
rebased on mater @cberner

cberner commented on presto-docs/src/main/sphinx/sql/explain-analyze.rst:
```diff
@@ -18,45 +18,55 @@ along with the cost of each operation.
 In the example below, you can see the CPU time spent in each stage, as well as the relative
 cost of each operator in the stage. Note that the relative cost of the operators is based on
-wall time, which may or may not be correlated to CPU time.
+wall time, which may or may not be correlated to CPU time. For each plan node you can see
```
I think the part about individual operators is out of date

sopel39 commented:
updated doc @cberner

cberner commented:
Merged, thanks! Sorry that took so long

kokosing commented:
Congrats!

pt., 27.01.2017, 18:12 użytkownik Christopher Berner <
notifications@github.com> napisał:

> Merged, thanks! Sorry that took so long
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/5499#issuecomment-275718956>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHN_--C4uM8Y_NYf7pN2AMBMlN7gn96fks5rWiVhgaJpZM4I7nrF>
> .
>

sopel39 commented:
Thanks! :D

ilfrin commented:
Yay! :D

pt., 27.01.2017, 20:39 użytkownik Karol Sobczak <notifications@github.com>
napisał:

> Thanks! :D
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/5499#issuecomment-275755070>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGF-yN3hdkc3BlMZGZJ4AvBpIDUTxF5lks5rWkf4gaJpZM4I7nrF>
> .
>
